db/backup.go

```diff
@@ -0,0 +1,211 @@
+	backupDb, err := sql.Open(Driver+"_custom", path)
+	if err != nil {
+		return err
+	}
+	defer backupDb.Close()
+
+	existingConn, err := d.writeDB.Conn(ctx)
+	if err != nil {
+		return err
+	}
+	defer existingConn.Close()
+
+	backupConn, err := backupDb.Conn(ctx)
+	if err != nil {
+		return err
+	}
+	defer backupConn.Close()
+
+	err = existingConn.Raw(func(existing any) error {
+		return backupConn.Raw(func(backup any) error {
+			var sourceOk, destOk bool
+			var sourceConn, destConn *sqlite3.SQLiteConn
+
+			if isBackup {
+				sourceConn, sourceOk = existing.(*sqlite3.SQLiteConn)
+				destConn, destOk = backup.(*sqlite3.SQLiteConn)
+			} else {
+				sourceConn, sourceOk = backup.(*sqlite3.SQLiteConn)
+				destConn, destOk = existing.(*sqlite3.SQLiteConn)
+			}
+
+			if !sourceOk {
+				return fmt.Errorf("error trying to convert source to sqlite connection")
+			}
+			if !destOk {
+				return fmt.Errorf("error trying to convert destination to sqlite connection")
+			}
+
+			backupOp, err := destConn.Backup("main", sourceConn, "main")
+			if err != nil {
+				return fmt.Errorf("error starting sqlite backup: %w", err)
+			}
+			defer backupOp.Close()
+
+			// Caution: -1 means that sqlite will hold a read lock until the operation finishes
+			// This will lock out other writes that could happen at the same time
+			done, err := backupOp.Step(-1)
+			if !done {
+				return fmt.Errorf("backup not done with step -1")
+			}
+			if err != nil {
+				return fmt.Errorf("error during backup step: %w", err)
+			}
+
+			err = backupOp.Finish()
+			if err != nil {
+				return fmt.Errorf("error finishing backup: %w", err)
+			}
+
+			return nil
+		})
+	})
+
+	if err == nil && !conf.Server.Backup.Bypass {
+		files, err := os.ReadDir(conf.Server.Backup.Path)
+		if err != nil {
+			return fmt.Errorf("unable to read database backup entries: %w", err)
+		}
+
+		times := []time.Time{}
+
+		for _, file := range files {
+			if !file.IsDir() {
+				submatch := backupRegex.FindStringSubmatch(file.Name())
+				if len(submatch) == 2 {
+					timestamp, err := time.Parse(time.RFC3339, submatch[1])
+					if err == nil {
+						times = append(times, timestamp)
+					}
+				}
+			}
+		}
+
+		slices.SortFunc(times, func(a, b time.Time) int {
+			return b.Compare(a)
+		})
+
+		toPrune := pruneBackups(ctx, times)
+
+		if len(toPrune) > 0 {
+			var errs []error
+
+			for _, time := range toPrune {
+				path := backupPath(&time)
+				err = os.Remove(path)
+
+				if err != nil {
+					errs = append(errs, err)
+				}
+			}
+
+			if len(errs) > 0 {
+				err = errors.Join(errs...)
+				log.Error(ctx, "Failed to delete one or more files", "errors", err)
+			}
+		}
+	}
+
+	return err
+}
+
+type backupSchedule struct {
+	count  int
+	format string
+	text   string
+}
+
+var (
+	backupRegex = regexp.MustCompile(backupRegexString)
+)
+
+func pruneBackups(ctx context.Context, times []time.Time) []time.Time {
```

Comment:
This is a really complex piece of logic (or I'm already too sleepy). Can we have some comments?